### ISSUE TITLE
How to deal with Master / Slave failover with pool

### ISSUE THREAD

**Issue by schleppy** [garyburd/redigo#150]
tl;dr: Is there an idiomatic way to deal with connections in a pool during a master failover situation where the existing master becomes a slave?

---

So I want to be able to deal with master / slave failover with redis instances.  The situation is like this: We have a redisEndpoint (DNS name) pointing to a master.  When our service starts it creates a pool of connections to redisEndpoint.  At any point the original redisEndpoint master could failover to a new master and itself become a slave.  

What I would like to do is identify a "READONLY" type error from redis and rather than call `conn.Close()` which will just put the connection back in the pool, I want to actually close the connection and decrement the active count (ala pool.release()) so the pool doesn't view the now closed connection as an outstanding resource from the pool.

I have done this in testing by adding a `Remove` method to the `Conn interface` and implementing dummy methods for `type conn` and `type errorConnection`.  The `Remove` method for `type pooledConnection` simply closes the underlying connection and calls `c.p.release()`.  This works but feels a bit clunky as it doesn't really make sense for the `type conn` to have a `Remove` method as it won't ever have any real meaning or functionality.

I thought it might be possible to have the `errorConnection.Close()` method check the connection for any error and if an error exists omit calling `put` on the pool and instead calling `release`.

**Comment by garyburd** [comment]
Use Redis 2.8.12 or later.  The server closes connections on configuration change [doc(http://redis.io/topics/sentinel-clients#sentinel-failover-disconnection).

**Comment by garyburd** [comment]
Please reopen if 2.8.12 does not solve the issue.
